### PR TITLE
enh(core): update script to manage gorgone configuration file

### PR DESCRIPTION
**Pull request by sc979**
# Pull Request Template

## Description

Generate the gorgoned.yml configuration file at 20.04.x upgrade.
Get the current user's configuration values to enable the legacy mode of the module

**Fixes** # (none)

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 2.8.x
- [ ] 18.10.x
- [ ] 19.04.x
- [ ] 19.10.x
- [x] 20.04.x (master)

<h2> How this pull request can be tested ? </h2>

1- default folders / OVF / ISO :
In the centreon's specific folder, the gorgoned.yml file should be created without throwing an error in the logs.
The default folder is "/etc/centreon".

2 - custom folders : 
You need to test the upgrade using custom folders, chosen in the fresh install form of a previous major release, and then upgrade to the 20.04.x.
These values should be backported in a new gorgoned.yml file.

#### Community contributors & Centreon team

- [x] I followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

#### Centreon team only

- [ ] I have made sure that the **unit tests** related to the story are successful.
- [ ] I have made sure that **unit tests cover 80%** of the code written for the story.
- [ ] I have made sure that **acceptance tests** related to the story are successful (**local and CI**)

### CODE REVIEW

File: www/install/php/Update-20.04.0-beta.1.php [hunk 1]
```diff
@@ -90,6 +89,111 @@
...
+            "",
+            fgets($file)
+        );
+        // checking for not commented string length
+        if (strpos($line, "#") === false && strlen($line) > 1) {
+            $line = explode("=", $line);
+            list($currentLine, $currentValue) = $line;
+
+            if (strlen($currentValue) > 1) {
+                switch ($currentLine) {
```
Comment (callapa) [inline www/install/php/Update-20.04.0-beta.1.php]: More simple: use an array to store the pattern values.
$patterns = ['instance_mode' => '/--INSTANCEMODE--/', ...]

File: www/install/php/Update-20.04.0-beta.1.php [hunk 2]
```diff
@@ -90,6 +89,110 @@
...
+        '/etc/snmp/centreon_traps',
+    ];
+
+    $instanceMode = "";
+
+    $fileToOpen = _CENTREON_ETC_ . './conf.pm';
+
+    // check if conf.pm exists and is not empty
+    if (!file_exists($fileToOpen) || 0 === filesize($fileToOpen)) {
+        throw new InvalidArgumentException('Missing \'conf.pm\' file');
```
Comment (callapa) [inline www/install/php/Update-20.04.0-beta.1.php]:
```suggestion
        throw new \InvalidArgumentException('Missing \'conf.pm\' file');
```

File: www/install/php/Update-20.04.0-beta.1.php (hunk 1 above)
Comment (sc979) [reply www/install/php/Update-20.04.0-beta.1.php]: The "instance_mode" isn't a key I need to keep.
The key need to be an integer matching the one of the second array to be able to use the preg_replace later

File: www/install/php/Update-20.04.0-beta.1.php [hunk 3]
```diff
@@ -90,6 +89,110 @@
...
+    while (!feof($file)) {
+        // removing useless characters
+        $line = str_replace(
+            [" ", ",", ";", "{", "}", "\"", "'", ">", "\n", "$"],
+            "",
+            fgets($file)
+        );
+        // checking for not commented string length
+        if (strpos($line, "#") === false && strlen($line) > 1) {
+            list($currentLine, $currentValue) = explode("=", $line);
```
[the remark this answers is not in the document]
Comment (sc979) [reply www/install/php/Update-20.04.0-beta.1.php]: I sanitized each lines to keep only "key=value" like sentences.
I'll try to find a way to check the returned state of the explode method

[the remark this answers is not in the document]
Comment (sc979) [reply www/install/php/Update-20.04.0-beta.1.php]: I added a check to avoid the case

File: www/install/php/Update-20.04.0-beta.1.php [hunk 4]
```diff
@@ -20,18 +20,104 @@
...
+            fgets($file)
+        );
+        // checking for not commented string length
+        if (strpos($line, "#") === false && strlen($line) > 1) {
+            $line = explode("=", $line);
+            isset($line[1]) ?: $line[1] = "";
+            list($currentLine, $currentValue) = $line;
+            if (strlen($currentValue) > 1) {
+                if (array_key_exists($currentLine, $patterns)) {
+                    var_dump($currentLine);
```
Comment (callapa) [inline www/install/php/Update-20.04.0-beta.1.php]:
```suggestion
```

File: www/install/php/Update-20.04.0-beta.1.php [hunk 5]
```diff
@@ -20,18 +20,104 @@
...
+                    var_dump($patterns[$currentLine]);
```
Comment (callapa) [inline www/install/php/Update-20.04.0-beta.1.php]:
```suggestion
```

File: www/install/php/Update-20.04.0-beta.1.php [hunk 6]
```diff
@@ -20,18 +20,102 @@
...
+        // removing useless characters
+        $line = str_replace(
+            [" ", ",", ";", "{", "}", "\"", "'", ">", "\n", "$"],
+            "",
+            fgets($file)
+        );
+        // checking for not commented string length
+        if (strpos($line, "#") === false && strlen($line) > 1) {
+            $line = explode("=", $line);
+            isset($line[1]) ?: $line[1] = "";
```
Comment (kduret) [inline www/install/php/Update-20.04.0-beta.1.php]:
```suggestion
        $line[1] = isset($line[1]) ?: "";
```

File: www/install/php/Update-20.04.0-beta.1.php [hunk 7]
```diff
@@ -20,18 +20,102 @@
...
+    }
+
+    // checking if the instance is a central or a poller and generating the configuration file
+    if ($instanceMode === 'central') {
+        $content = file_get_contents('../var/configFileGorgoneTemplate');
+        $content = preg_replace($pattern, $userValues, $content);
+        file_put_contents(_CENTREON_ETC_ . '/gorgoned.yml', $content);
+    }
+    if (!file_exists(_CENTREON_ETC_ . '/gorgoned.yml') || 0 === filesize(_CENTREON_ETC_ . '/gorgoned.yml')) {
+        throw new \InvalidArgumentException('Gorgone configuration file wasn\'t created properly');
```
Comment (kduret) [inline www/install/php/Update-20.04.0-beta.1.php]:
```suggestion
        throw new \InvalidArgumentException('Gorgone configuration file isn\'t created properly');
```

File: www/install/php/Update-20.04.0-beta.1.php [hunk 8]
```diff
@@ -20,18 +20,102 @@
...
+        $line = str_replace(
+            [" ", ",", ";", "{", "}", "\"", "'", ">", "\n", "$"],
+            "",
+            fgets($file)
+        );
+        // checking for not commented string length
+        if (strpos($line, "#") === false && strlen($line) > 1) {
+            $line = explode("=", $line);
+            isset($line[1]) ?: $line[1] = "";
+            list($currentLine, $currentValue) = $line;
```
[the remark this answers is not in the document]
Comment (sc979) [reply www/install/php/Update-20.04.0-beta.1.php]: no it'll be affected an empty value ("") -> line 75

File: lang/es_ES.UTF-8/LC_MESSAGES/messages.po [hunk 9]
```diff
@@ -3231,14 +3231,19 @@ msgid "You seem to be having trouble with your upgrade."
...
 
 #: centreon-web/www/install/step_upgrade/step4.php:71
 #, php-format
 msgid "The SQL files are located in %s"
 msgstr "Los archivos SQL se encuentran en %s"
 
+#: centreon-web/www/install/step_upgrade/step4.php:87
+#, php-format
+msgid "Please check the \"upgrade.log\" and the \"sql-error.log\" located in \"%s\" for more details"
+msgstr ""
```
Comment (kduret) [inline lang/es_ES.UTF-8/LC_MESSAGES/messages.po]: if you put it empty, es translation will display empty.
If you cannot translate it, just remove those lines

File: lang/pt_BR.UTF-8/LC_MESSAGES/messages.po [hunk 10]
```diff
@@ -11996,14 +11996,19 @@ msgid "You seem to be having trouble with your upgrade."
...
 
 #: centreon-web/www/install/step_upgrade/step4.php:71
 #, php-format
 msgid "The SQL files are located in %s"
 msgstr "Os arquivos SQL estão localizados em %s"
 
+#: centreon-web/www/install/step_upgrade/step4.php:87
+#, php-format
+msgid "Please check the \"upgrade.log\" and the \"sql-error.log\" located in \"%s\" for more details"
+msgstr ""
```
Comment (kduret) [inline lang/pt_BR.UTF-8/LC_MESSAGES/messages.po]: if you put it empty, br translation will display empty.
If you cannot translate it, just remove those lines

File: lang/pt_PT.UTF-8/LC_MESSAGES/messages.po [hunk 11]
```diff
@@ -11997,14 +11997,19 @@ msgid "You seem to be having trouble with your upgrade."
...
 
 #: centreon-web/www/install/step_upgrade/step4.php:71
 #, php-format
 msgid "The SQL files are located in %s"
 msgstr "Os ficheiros SQL estão localizados em %s"
 
+#: centreon-web/www/install/step_upgrade/step4.php:87
+#, php-format
+msgid "Please check the \"upgrade.log\" and the \"sql-error.log\" located in \"%s\" for more details"
+msgstr ""
```
Comment (kduret) [inline lang/pt_PT.UTF-8/LC_MESSAGES/messages.po]: if you put it empty, pt translation will display empty.
If you cannot translate it, just remove those lines

File: lang/es_ES.UTF-8/LC_MESSAGES/messages.po [hunk 12]
```diff
@@ -3231,14 +3231,19 @@ msgid "You seem to be having trouble with your upgrade."
...
 msgstr "Puede consultar la línea que es problemática para obtener más información sobre el incidente."
 
 #: centreon-web/www/install/step_upgrade/step4.php:71
 #, php-format
 msgid "The SQL files are located in %s"
 msgstr "Los archivos SQL se encuentran en %s"
 
+#: centreon-web/www/install/step_upgrade/step4.php:87
+#, php-format
+msgid "Please check the \"upgrade.log\" and the \"sql-error.log\" located in \"%s\" for more details"
```
Comment (sc979) [inline lang/es_ES.UTF-8/LC_MESSAGES/messages.po]:
```suggestion
# msgid "Please check the \"upgrade.log\" and the \"sql-error.log\" located in \"%s\" for more details"
```

File: lang/es_ES.UTF-8/LC_MESSAGES/messages.po (hunk 9 above)
Comment (sc979) [inline lang/es_ES.UTF-8/LC_MESSAGES/messages.po]:
```suggestion
# msgstr ""
```

File: lang/pt_BR.UTF-8/LC_MESSAGES/messages.po [hunk 13]
```diff
@@ -11996,14 +11996,19 @@ msgid "You seem to be having trouble with your upgrade."
...
 msgstr "Você deve verificar a linha que esta causando  o problema para poder-mos encontrar uma resolução para este caso."
 
 #: centreon-web/www/install/step_upgrade/step4.php:71
 #, php-format
 msgid "The SQL files are located in %s"
 msgstr "Os arquivos SQL estão localizados em %s"
 
+#: centreon-web/www/install/step_upgrade/step4.php:87
+#, php-format
+msgid "Please check the \"upgrade.log\" and the \"sql-error.log\" located in \"%s\" for more details"
```
Comment (sc979) [inline lang/pt_BR.UTF-8/LC_MESSAGES/messages.po]:
```suggestion
#~ msgid "Please check the \"upgrade.log\" and the \"sql-error.log\" located in \"%s\" for more details"
```

**Pull request merged**